### PR TITLE
add allow_primitive_to_str and allow_decimal_to_str instead of allow_number_to_str

### PR DESCRIPTION
**Pull request by Modexus**
PR for #6805

### CODE REVIEW

**Comment by Modexus** [comment]
@mariosasko pytest seems to be missing on windows?

**Comment by mariosasko** [comment]
CI is not behaving well today 🙂

File: src/datasets/table.py [hunk 1]
```diff
@@ -1933,13 +1940,14 @@ def array_cast(array: pa.Array, pa_type: pa.DataType, allow_number_to_str=True):
...
-            and (pa.types.is_floating(array.type) or pa.types.is_integer(array.type))
+        if pa.types.is_string(pa_type) and (
+            (not allow_primitive_to_str and pa.types.is_primitive(array.type))
+            or (not allow_decimal_to_str and pa.types.is_decimal(array.type))
         ):
             raise TypeError(
-                f"Couldn't cast array of type {array.type} to {pa_type} since allow_number_to_str is set to {allow_number_to_str}"
+                f"Couldn't cast array of type {array.type} to {pa_type} "
+                f"since allow_primitive_to_str is set to {allow_primitive_to_str} "
+                f"and allow_decimal_to_str is set to {allow_decimal_to_str}."
```
Comment (mariosasko) [inline src/datasets/table.py:1950]: Maybe let's be more precise here depending on the type - if it's primitive, then we should mention `allow_primitive_to_str `, and if it's decimal  `allow_decimal_to_str `

**Review by mariosasko — COMMENTED** [review]
Thanks! A few comments:

File: src/datasets/arrow_writer.py [hunk 2]
```diff
@@ -241,7 +243,7 @@ def __arrow_array__(self, type: Optional[pa.DataType] = None):
                             cast_to_python_objects(data, only_1d_for_numpy=True, optimize_list_casting=False)
                         )
                         if type is not None:
-                            out = cast_array_to_feature(out, type, allow_number_to_str=True)
+                            out = cast_array_to_feature(out, type, allow_primitive_to_str=True)
```
[the remark this answers is not in the document]
Comment (Modexus) [reply src/datasets/arrow_writer.py]: Didn't think about the other calls..., I added this now.

File: src/datasets/table.py (hunk 1 above)
Comment (Modexus) [reply src/datasets/table.py]: Split this into two separate errors.

**Review by mariosasko — APPROVED** [review]
LGTM! Also, thanks for adding tests!

**Comment by mariosasko** [comment]
I could find an instance of the `allow_number_to_str` parameter (or `array_cast`/`cast_array_to_feature` more generally) being used in the wild. So, I think simply removing `allow_number_to_str` instead of deprecating it should be fine, considering `array_cast`/`cast_array_to_feature` are somewhat hidden. Do you agree @lhoestq?

**Comment by lhoestq** [comment]
Yup we can remove without any deprecation cycle

**Review by lhoestq — APPROVED** [review]
Nice fix !

**Pull request merged**